ocean_lib/ocean/v4/ocean_assets.py

```diff
@@ -0,0 +1,200 @@
+#
+# Copyright 2021 Ocean Protocol Foundation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""Ocean module."""
+import copy
+import logging
+import os
+from typing import Optional, Type, Tuple
+
+from eth_utils import remove_0x_prefix
+from web3 import Web3
+
+from ocean_lib.agreements.service_types import ServiceTypesV4
+from ocean_lib.assets.v4.asset import V4Asset
+from ocean_lib.config import Config
+from ocean_lib.data_provider.data_service_provider import DataServiceProvider
+from ocean_lib.exceptions import ContractNotFound, AquariusError
+from ocean_lib.models.v4.erc721_factory import ERC721FactoryContract
+from ocean_lib.models.v4.erc721_token import ERC721Token
+from ocean_lib.services.v4.service import V4Service
+from enforce_typing import enforce_types
+
+from ocean_lib.web3_internal.constants import ZERO_ADDRESS
+from ocean_lib.web3_internal.wallet import Wallet
+from tests.resources.helper_functions import get_address_of_type
+
+logger = logging.getLogger("ocean")
+
+
+@enforce_types
+class OceanAssetV4:
+    """Ocean asset class for V4."""
+
+    def __init__(
+        self,
+        config: Config,
+        web3: Web3,
+        data_provider: Type[DataServiceProvider],
+        ddo_registry_address: str,
+    ) -> None:
+        """Initialises OceanAssets object."""
+        self._config = config
+        self._web3 = web3
+        self._metadata_cache_uri = config.metadata_cache_uri
+        self._data_provider = data_provider
+        self._metadata_registry_address = ddo_registry_address
+
+        downloads_path = os.path.join(os.getcwd(), "downloads")
+        if self._config.has_option("resources", "downloads.path"):
+            downloads_path = (
+                self._config.get("resources", "downloads.path") or downloads_path
+            )
+        self._downloads_path = downloads_path
+
+    def validate(self, metadata: dict) -> Tuple[bool, list]:
+        """
+        Validate that the metadata is ok to be stored in aquarius.
+
+        :param metadata: dict conforming to the Metadata accepted by Ocean Protocol.
+        :return: (bool, list) list of errors, empty if valid
+        """
+        return self._get_aquarius(self._metadata_cache_uri).validate_metadata(metadata)
+
+    def _add_defaults(
+        self, services: list, data_token: str, files: str, provider_uri: str
+    ) -> list:
+
+        has_access_service = False
+        for service in services:
```

Comment:
please use a comprehension

Reply:
Instead using a list comprehension and after a filter, I prefer using `any(map(lambda s: s.type == ServiceTypesV4.ASSET_ACCESS, services))`. Is it ok like that?